composer/callbacks/checkpoint_saver.py

```diff
@@ -397,17 +397,32 @@ def _save_checkpoint(self, state: State, logger: Logger):
                 os.remove(symlink)
             except FileNotFoundError:
                 pass
-            os.symlink(os.path.relpath(filename, os.path.dirname(symlink)), symlink)
+            os.symlink(os.path.relpath(saved_path, os.path.dirname(symlink)), symlink)
 
         # if remote file name provided, upload the checkpoint
         if self.remote_file_name is not None:
-            remote_file_name = self.remote_file_name.format(
-                state,
-                is_deepspeed,
-            ).lstrip('/')
+            # Use the stem of remote_filename (with rank info stripped) as a prefix folder.
+            # e.g. s3://my-bucket/path/to/checkpoints/ep1-ba2-rank3.pt is now s3://my-bucket/path/to/checkpoints/ep1-ba2/ep1-ba2-rank3.pt
+            if state.fsdp_sharded_state_dict_enabled and state.sharded_ckpt_prefix_dir is not None:
```

Comment:
What happens if `state.fsdp_sharded_state_dict_enabled` and `state.sharded_ckpt_prefix_dir is None`? Is that possible?

Reply:
nope just trying to please pyright

Reply:
well unless the user explicitly sets it to none

Reply:
could you do an assert for the invariant rather than adding it to the if statement? I think that would be more clear.